torch_tvm/register.cpp

```diff
@@ -22,55 +23,93 @@ static std::string device_type = "cpu";
+                },
+                options)});
 
   // Register the pass that fuses parts of the graph into
   // a tvm::CompilationGroup
-  RegisterPass pass([tvm_sym](std::shared_ptr<Graph>& g) {
+  RegisterPass pass([tvm_sym](std::shared_ptr<Graph> &g) {
     if (fusion_enabled) {
       CustomFuseGraph(g, isSupported, tvm_sym);
     }
   });
 
   // python API to enable and disable tvm fusion
-  m.def(
-      "enable",
-      [](int opt_level_,
-         bool strict_,
-         std::string device_type_,
-         std::string device_,
-         std::string host_) {
-        fusion_enabled = true;
-        strict = strict_;
-        opt_level = opt_level_;
-        device_type = device_type_;
-        device = device_;
-        host = host_;
-      },
-      py::arg("opt_level") = 2,
-      py::arg("strict") = false,
-      py::arg("device_type") = "cpu",
-      py::arg("device") = "llvm -mcpu=core-avx2",
-      py::arg("host") = "llvm -mcpu=core-avx2");
+  m.def("enable",
+        [](int opt_level_, bool strict_, std::string device_type_,
+           std::string device_, std::string host_) {
+          fusion_enabled = true;
+          strict = strict_;
+          opt_level = opt_level_;
+          device_type = device_type_;
+          device = device_;
+          host = host_;
+        },
+        py::arg("opt_level") = 2, py::arg("strict") = false,
+        py::arg("device_type") = "cpu",
+        py::arg("device") = "llvm -mcpu=core-avx2",
+        py::arg("host") = "llvm -mcpu=core-avx2");
 
   m.def("disable", []() { fusion_enabled = false; });
+  m.def("register_operator", [tvm_sym](std::string torch_name,
+                                       std::string relay_name) {
+    RegisterTVMOperator reg_tvm_operator(
+        {{Symbol::fromQualString("tvm::" + torch_name),
+          [relay_name](Node *node, tvm::Array<tvm::relay::Expr> inputs) {
+            auto op = tvm::relay::Op::Get(relay_name);
+            AT_ASSERT(op);
+            return tvm::relay::CallNode::make(op, inputs, tvm::Attrs(), {});
+          }}});
+
+    auto relay_op = tvm::relay::Op::Get(relay_name);
+    AT_ASSERT(relay_op);
+    std::vector<Argument> torch_inputs;
+    Graph wrapper_graph;
+    std::vector<Value *> graph_inputs;
+    for (auto i = 0; i < relay_op->num_inputs; ++i) {
+      torch_inputs.emplace_back();
+      graph_inputs.emplace_back(wrapper_graph.addInput());
+    }
+    auto sym = Symbol::fromQualString("tvm::" + torch_name);
+    Node *node = wrapper_graph.create(sym, graph_inputs);
+    wrapper_graph.appendNode(node);
+    AT_CHECK(node->outputs().size() == 1,
+             "Currently only single output relay ops "
+             "can be custom registered from Python.");
+    wrapper_graph.registerOutput(node->output());
+    // Convert node to a tvm compilation group containing the node
+    node = SubgraphUtils::createSingletonSubgraph(node, tvm_sym);
+    auto cc = std::make_shared<TVMCompiler>(node, opt_level, strict,
```

Comment:
from what I observed here, for each custom operators, we are creating a single separate `TVMCompiler` group and execute them separately? will the `customFusion` pass take the compilation group and merge it with the bigger one?

Reply:
customFusion will never be run on these operators because this interface is for eager mode only